In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv('./lending_data.csv')

# Review the DataFrame
lending_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df[lending_df.columns[:7]]

In [4]:
# Review the y variable Series
y

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64

In [5]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [21]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X,
                                                   y,
                                                   random_state=1,
                                                   stratify=y)


print(f'Number of X_train: {X_train.shape[0]}\nNumber of y_train: {y_train.shape[0]}')

print(f'Number of X_test: {X_test.shape[0]}\nNumber of y_test: {y_test.shape[0]}')

Number of X_train: 58152
Number of y_train: 58152
Number of X_test: 19384
Number of y_test: 19384


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=1)


# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
testing_predictions =  lr_model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [28]:
# Print the balanced_accuracy score of the model
balanced_accuracy = balanced_accuracy_score(y_test, testing_predictions)
balanced_accuracy

0.9442676901753825

In [29]:
# Generate a confusion matrix for the model
test_matrix = confusion_matrix(y_test, testing_predictions)
test_matrix

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [30]:
# Print the classification report for the model

# *!*!*!*!*NOTE: More significant digits were added 'digits=5' to the classification report to expose the presence of
# false positive and false negatives in the precision and recall of the 'Healthy Loan (0)'
testing_report = classification_report(y_test, testing_predictions, target_names=['Healthy Loan (0)', 'High-Risk Loan (1)'], digits=5)
print(testing_report)

                    precision    recall  f1-score   support

  Healthy Loan (0)    0.99643   0.99574   0.99608     18759
High-Risk Loan (1)    0.87461   0.89280   0.88361       625

          accuracy                        0.99242     19384
         macro avg    0.93552   0.94427   0.93985     19384
      weighted avg    0.99250   0.99242   0.99245     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** <br>The logistic regression model predicts the '0' (healthy loan) ALMOST perfectly with a:<br> 
<ul> 
    <li>precision = 0.99643</li>
    <li>recall = 0.99574</li>
    <li>F1 Score = 0.99608</li>
</ul>
Whereas, the logistic regression model predicts the '1' (high-risk loan) with significant inaccuracy with a:<br>
<ul>
    <li>precision = 0.87461</li>
    <li>recall = 0.89280</li>
    <li>F1 Score = 0.88361</li>
</ul>
And the overall (balanced) accuracy is 0.94427 (which is the same as the macro average recall).

### False Predictions
Let's look at the false positives and false negatives.<br>
#### False POSITIVES
Here, there was 80 / (80 + 558) = 1.00 - precision(1) = 0.12539 or 12.54% false POSITIVES of the predicted positives which in this context means there were 12.54% of actual 'healthy loans' being mislabeled as 'high-risk loans' and so, potential lost profits for not giving loans to those accounts. 

#### False NEGATIVES
Here, there was 67 / (67 + 18679) = 1.00 - precision(0) = 0.00357 or 0.357% false NEGATIVES of the predicted negatives which in this context means there were 0.357% of actual 'high-risk loans' being mislabeled as 'healthy loans' and so, potential lost profits may occur from giving the loans to those accounts. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros  = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resample_train, y_resample_train = ros.fit_resample(X_train, y_train)

In [14]:
# Count the distinct values of the resampled labels data
y_resample_train.value_counts()

0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [15]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model2 = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
lr_model2 = logistic_regression_model2.fit(X_resample_train, y_resample_train)

# Make a prediction using the testing data
testing_predictions2 =  lr_model2.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Print the balanced_accuracy score of the model 
balanced_accuracy2 = balanced_accuracy_score(y_test, testing_predictions2)
balanced_accuracy2

0.9959744975744975

In [17]:
# Generate a confusion matrix for the model
test_matrix2 = confusion_matrix(y_test, testing_predictions2)
test_matrix2

array([[18668,    91],
       [    2,   623]], dtype=int64)

In [25]:
# Print the classification report for the model
# *!*!*!*!*NOTE: More significant digits were added 'digits=5' to the classification report to expose the presence of
# false positive and false negatives in the precision and recall of the 'Healthy Loan (0)'
testing_report2 = classification_report(y_test, testing_predictions2, target_names=['Healthy Loan (0)', 'High-Risk Loan (1)'], digits=5)
print(testing_report2)

                    precision    recall  f1-score   support

  Healthy Loan (0)    0.99989   0.99515   0.99752     18759
High-Risk Loan (1)    0.87255   0.99680   0.93055       625

          accuracy                        0.99520     19384
         macro avg    0.93622   0.99597   0.96403     19384
      weighted avg    0.99579   0.99520   0.99536     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**   <br>The logistic regression model predicts the '0' (healthy loan) ALMOST perfectly with a:<br> 
<ul> 
    <li>precision = 0.99989</li>
    <li>recall = 0.99515</li>
    <li>F1 Score = 0.99751</li>
</ul>
Whereas, the logistic regression model predicts the '1' (high-risk loan) with some inaccuracy with a:<br>
<ul>
    <li>precision = 0.87255</li>
    <li>recall = 0.99680</li>
    <li>F1 Score = 0.93204</li>
</ul>
And the overall (balanced) accuracy is 0.99597 (which is the same as the macro average recall).

### False Predictions
Let's look at the false positives and false negatives.<br>

#### False POSITIVES
Here, there was 91 / (91 + 623) = 1.00 - precision(1) = 0.12745 or 12.75% false POSITIVES of the predicted positives which in this context means there were 12.75% of actual 'healthy loans' being mislabeled as 'high-risk loans' and so, potential lost profits for not giving loans to those accounts. 

#### False NEGATIVES
Here, there was 2 / (2 + 18668) = 1.00 - precision(0) = 0.00011 or 0.001% false NEGATIVES of the predicted negatives which in this context means there were 0.001% of actual 'high-risk loans' being mislabeled as 'healthy loans' and so, potential lost profits may occur from giving the loans to those accounts. 

#### Assessment
So, we see that the oversampled model is dramatically better at avoiding false negatives-- only 0.001% of the predicted negatives are false negatives for the oversampled model whereas the single-sampled model (original model) returns 0.357% false negatives! That's  a RELATIVE difference of 35,700%!!